ValheimPlus/Inventory.cs

```diff
@@ -1,19 +1,132 @@
-﻿using HarmonyLib;
-using System;
-using ValheimPlus.Configurations;
-
-namespace ValheimPlus
-{
-    [HarmonyPatch(typeof(Inventory), "IsTeleportable")]
-    public static class noItemTeleportPrevention
-    {
-        private static void Postfix(ref Boolean __result)
-        {
-            if (Configuration.Current.Items.IsEnabled)
-            {
-                if (Configuration.Current.Items.noTeleportPrevention)
-                    __result = true;
-            }
-        }
-    }
-}
+﻿using HarmonyLib;
+using System;
+using UnityEngine;
+using UnityEngine.UI;
+using ValheimPlus.Configurations;
+
+namespace ValheimPlus
+{
+    [HarmonyPatch(typeof(Inventory), "IsTeleportable")]
+    public static class noItemTeleportPrevention
+    {
+        private static void Postfix(ref Boolean __result)
+        {
+            if (Configuration.Current.Items.IsEnabled)
+            {
+                if (Configuration.Current.Items.noTeleportPrevention)
+                    __result = true;
+            }
+        }
+    }
+
+    [HarmonyPatch(typeof(Inventory), "TopFirst", new Type[] { typeof(ItemDrop.ItemData) })]
+    public static class changeInventoryFillOrder
+    {
+        public static bool Prefix(ref Boolean __result)
+        {
+            if (Configuration.Current.Inventory.IsEnabled &&
+                Configuration.Current.Inventory.inventoryFillTopToBottom)
+            {
+                __result = true;
+                return false;
+            }
+            else
+            {
+                return true;
+            }
+        }
+    }
+
+    [HarmonyPatch(typeof(Inventory), MethodType.Constructor)]
+    [HarmonyPatch(new Type[] { typeof(string), typeof(Sprite), typeof(int), typeof(int) })]
+    public static class changeInventorySlotCount
+    {
+        static void Prefix(ref string name, ref Sprite bkg, ref int w, ref int h)
+        {
+            if (Configuration.Current.Inventory.IsEnabled)
+            {
+                // Player inventory number of rows, clamp value in range 4-20
+                int playerInventoryRows = Math.Min(20, Math.Max(4, Configuration.Current.Inventory.playerInventoryRows));
+
+                // Wood chest number of columns, clamp value in range 5-8
+                int woodChestColumns = Math.Min(8, Math.Max(5, Configuration.Current.Inventory.woodChestColumns));
+
+                // Wood chest number of rows, clamp value in range 2-10
+                int woodChestRows = Math.Min(10, Math.Max(2, Configuration.Current.Inventory.woodChestRows));
+
+                // Iron chest number of columns, clamp value in range 6-8
+                int ironChestColumns = Math.Min(8, Math.Max(6, Configuration.Current.Inventory.ironChestColumns));
+
+                // Iron chest number of rows, clamp value in range 3-20
+                int ironChestRows = Math.Min(20, Math.Max(3, Configuration.Current.Inventory.ironChestRows));
+
+                // Wood chest
+                if (h == 2 && w == 5)

```

Comment:
@Zogniton can we check on name equality instead of h & w? That way we can separate out iron chest from boat and cart inventory. [Check out how we do this in Fireplace.cs for convention.](https://github.com/valheimPlus/ValheimPlus/blob/84341c51fae2e9fd3e24428ef6ca561544d0b1f0/ValheimPlus/Fireplace.cs#L50)

Reply:
I tried this initially and it could probably be done, there were 2 things i did not like about it:
1. Chest names seem to be inconsistent, sometimes you get "Chest" and "Inventory", other times you get $piece_chestwood
2. I felt that adding 4 more parameters in the config file is a bit much considering vanilla iron chests, cart and boats are the same size

I can still try to make this work if needed

Reply:
Hmm I'll look more into this